ambry-store/src/main/java/com/github/ambry/store/BlobStoreCompactor.java

```diff
@@ -1467,7 +1485,16 @@ public boolean alreadyExists(PersistentIndex idx, FileSpan searchSpan, StoreKey
                 "Delete's latest state can't be put for key" + currentKey + " in store " + dataDir);
           }
           if (currentLatestState.isDelete() && currentLatestState.getLifeVersion() == currentValue.getLifeVersion()) {
-            validEntries.add(entry);
+            // check if this is a removable delete tombstone.
+            IndexValue putValue = getPutValueFromSrc(currentKey, currentValue, indexSegment);
+            if (putValue == null && config.storeCompactionPurgeExpiredDeleteTombstone && isDeleteTombstoneRemovable(
+                currentValue)) {
+              logger.debug(
+                  "Delete tombstone of {} (with expiration time {} ms) is removable and won't be copied to target log segment",
+                  currentKey, currentValue.getExpiresAtMs());
```

Comment:
nit: we can reorganize this block a bit to avoid unnecessary code execution.
```
if (config.storeCompactionPurgeExpiredDeleteTombstone && isDeleteTombstoneRemovable(currentValue) && getPutValueFromSrc(currentKey,  currentValue, indexSegment) == null) {
    logger.debug("xxxx");
} else {
}
```

Reply:
yep, your suggestion is better